File: lib/pdfParser.ts
import { exec } from 'child_process'
import { promisify } from 'util'
import fs from 'fs/promises'
import path from 'path'

const execAsync = promisify(exec)

export interface PdfPageText {
  pageNumber: number
  text: string
}

export interface PdfParseResult {
  pageCount: number
  pages: PdfPageText[]
}

/**
 * 解析 PDF 文件，提取每一页的文本
 * 使用 Python pdfplumber 库，文本提取质量更好
 */
export async function parsePdfFile(filePath: string): Promise<PdfParseResult> {
  try {
    const pythonPath = path.join(process.cwd(), 'venv', 'bin', 'python3')
    const scriptPath = path.join(process.cwd(), 'lib', 'pdf_parser.py')

    // 使用 Python 脚本解析 PDF
    const { stdout, stderr } = await execAsync(`"${pythonPath}" "${scriptPath}" "${filePath}"`, {
      maxBuffer: 50 * 1024 * 1024 // 50MB buffer for large PDFs
    })

    if (stderr) {
      console.warn('Python PDF 解析警告:', stderr)
    }

    const result = JSON.parse(stdout)

    if (!result.success) {
      throw new Error(result.error || 'PDF 解析失败')
    }

    return {
      pageCount: result.pageCount,
      pages: result.pages
    }
  } catch (error) {
    console.error('PDF 解析失败:', error)
    throw new Error(`PDF 解析失败: ${error instanceof Error ? error.message : '未知错误'}`)
  }
}

/**
 * 简化版本：使用相同的 Python 解析器
 */
export async function parsePdfFileSimple(filePath: string): Promise<PdfParseResult> {
  return parsePdfFile(filePath)
}

/**
 * 生成唯一的文件名
 */
export function generateUniqueFilename(originalName: string): string {
  const timestamp = Date.now()
  const random = Math.random().toString(36).substring(2, 8)
  const ext = path.extname(originalName)
  const nameWithoutExt = path.basename(originalName, ext)
  return `${nameWithoutExt}-${timestamp}-${random}${ext}`
}

File: lib/pdf_parser.py
#!/usr/bin/env python3
"""
PDF 解析服务 - 使用 PyPDF2 或 pdfplumber
"""

import sys
import json
from pathlib import Path

try:
    import pdfplumber
    USE_PDFPLUMBER = True
except ImportError:
    USE_PDFPLUMBER = False
    try:
        from PyPDF2 import PdfReader
    except ImportError:
        print(json.dumps({
            "success": False,
            "error": "请安装 pdfplumber 或 PyPDF2: pip install pdfplumber"
        }))
        sys.exit(1)


def parse_pdf_with_pdfplumber(file_path):
    """使用 pdfplumber 解析 PDF（推荐，文本提取质量更好）"""
    pages = []

    with pdfplumber.open(file_path) as pdf:
        page_count = len(pdf.pages)

        for i, page in enumerate(pdf.pages, start=1):
            text = page.extract_text() or ""
            pages.append({
                "pageNumber": i,
                "text": text.strip()
            })

    return page_count, pages


def parse_pdf_with_pypdf2(file_path):
    """使用 PyPDF2 解析 PDF（备用方案）"""
    pages = []

    reader = PdfReader(file_path)
    page_count = len(reader.pages)

    for i, page in enumerate(reader.pages, start=1):
        text = page.extract_text() or ""
        pages.append({
            "pageNumber": i,
            "text": text.strip()
        })

    return page_count, pages


def main():
    if len(sys.argv) < 2:
        print(json.dumps({
            "success": False,
            "error": "Usage: python3 pdf_parser.py <pdf_file_path>"
        }))
        sys.exit(1)

    file_path = sys.argv[1]

    if not Path(file_path).exists():
        print(json.dumps({
            "success": False,
            "error": f"File not found: {file_path}"
        }))
        sys.exit(1)

    try:
        if USE_PDFPLUMBER:
            page_count, pages = parse_pdf_with_pdfplumber(file_path)
        else:
            page_count, pages = parse_pdf_with_pypdf2(file_path)

        result = {
            "success": True,
            "pageCount": page_count,
            "pages": pages
        }

        print(json.dumps(result, ensure_ascii=False))

    except Exception as e:
        print(json.dumps({
            "success": False,
            "error": str(e)
        }))
        sys.exit(1)


if __name__ == "__main__":
    main()
